File: netlify/functions/lib/connectors/fmp.ts
/**
 * Financial Modeling Prep fundamentals connector.
 * Ported from backend/connectors/fmp.py. Optional — the API key lives in
 * system_settings.fmp_api_key (falling back to the FMP_API_KEY env var for
 * the GitHub Actions runner). No key => no-op, and the scoring engine
 * degrades those sub-scores to neutral 50. Capped at 30 symbols/run to
 * respect the free tier (250 req/day; 3 calls/symbol).
 */
import { db } from "../db.js";

const BASE = "https://financialmodelingprep.com/api/v3";

export interface FmpFundamental {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ratios_ttm: Record<string, any>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  metrics_ttm: Record<string, any>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  estimates: any[];
}

async function getJson(path: string, apiKey: string, params: Record<string, string> = {}): Promise<unknown> {
  const qs = new URLSearchParams({ ...params, apikey: apiKey }).toString();
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 20000);
  const r = await fetch(`${BASE}/${path.replace(/^\//, "")}?${qs}`, { signal: controller.signal });
  clearTimeout(timer);
  if (r.status !== 200) {
    const body = (await r.text()).slice(0, 120);
    throw new Error(`FMP ${path} HTTP ${r.status}: ${body}`);
  }
  return r.json();
}

async function resolveApiKey(): Promise<string> {
  const { data } = await db.from("system_settings").select("value").eq("key", "fmp_api_key").maybeSingle();
  if (data?.value) return String(data.value);
  return process.env.FMP_API_KEY || "";
}

/** Fetch ratios-ttm / key-metrics-ttm / analyst-estimates per symbol (capped
 * at 30). Returns {} when no key is configured. Per-symbol failures are
 * skipped (fault-isolated), matching the Python connector. */
export async function fetchFmpFundamentals(symbols: string[]): Promise<Record<string, FmpFundamental>> {
  const apiKey = await resolveApiKey();
  if (!apiKey) return {};

  const out: Record<string, FmpFundamental> = {};
  for (const sym of symbols.slice(0, 30)) {
    const ticker = `${sym}.NS`;
    try {
      const ratios = await getJson(`ratios-ttm/${ticker}`, apiKey);
      const metrics = await getJson(`key-metrics-ttm/${ticker}`, apiKey);
      const est = await getJson(`analyst-estimates/${ticker}`, apiKey, { limit: "4" });
      out[sym] = {
        ratios_ttm: Array.isArray(ratios) ? ratios[0] || {} : (ratios as Record<string, unknown>) || {},
        metrics_ttm: Array.isArray(metrics) ? metrics[0] || {} : (metrics as Record<string, unknown>) || {},
        estimates: Array.isArray(est) ? est : [],
      };
    } catch (err) {
      console.warn(`fmp ${sym} failed:`, err instanceof Error ? err.message : err);
    }
  }
  return out;
}

File: legacy/python-fastapi/backend/connectors/fmp.py
"""Financial Modeling Prep (secondary fundamentals). Requires free API key (250/day).

Key is optional — stored in system_settings under `fmp_api_key`. When absent, the
connector is a no-op and the report just falls back to yfinance.info.
"""
from __future__ import annotations
import logging
from typing import Any, Dict, List, Optional

import httpx

from connectors.base import BaseConnector
from services.settings import get as _get_setting

logger = logging.getLogger(__name__)

BASE = "https://financialmodelingprep.com/api/v3"


async def _get_json(path: str, api_key: str, params: Optional[Dict[str, Any]] = None) -> Any:
    params = {**(params or {}), "apikey": api_key}
    async with httpx.AsyncClient(timeout=20) as client:
        r = await client.get(f"{BASE}/{path.lstrip('/')}", params=params)
    if r.status_code != 200:
        raise RuntimeError(f"FMP {path} HTTP {r.status_code}: {r.text[:120]}")
    return r.json()


class FMPConnector(BaseConnector):
    """Enriches fundamentals for NSE symbols.

    FMP uses the `.NS` suffix for NSE tickers (same as yfinance).
    """
    name = "fmp_fundamentals"
    category = "fundamentals"
    max_retries = 2

    async def fetch(self, symbols: List[str]) -> Dict[str, Any]:
        api_key = await _get_setting("fmp_api_key", "")
        if not api_key:
            return {"skipped": "no_api_key", "data": {}}

        out: Dict[str, Any] = {}
        # FMP accepts comma-separated tickers on some endpoints but profile/ratios are per-symbol.
        for sym in symbols[:30]:  # cap to preserve free quota
            ticker = f"{sym}.NS"
            try:
                # Ratios TTM — single most useful endpoint: ROE, D/E, interest coverage, margins, etc.
                ratios = await _get_json(f"ratios-ttm/{ticker}", api_key)
                # Key metrics TTM — EV/EBITDA, FCF yield, ROIC
                metrics = await _get_json(f"key-metrics-ttm/{ticker}", api_key)
                # Analyst estimates (next quarter)
                est = await _get_json(f"analyst-estimates/{ticker}", api_key, {"limit": 4})
                out[sym] = {
                    "ratios_ttm": (ratios or [{}])[0] if isinstance(ratios, list) else (ratios or {}),
                    "metrics_ttm": (metrics or [{}])[0] if isinstance(metrics, list) else (metrics or {}),
                    "estimates": est or [],
                }
            except Exception as e:  # noqa: BLE001
                logger.warning("FMP %s failed: %s", sym, e)
        return {"skipped": None, "data": out}
